File: src/main/java/dev/jlo/gamemodes/paper/SafetyPolicy.java
package dev.jlo.gamemodes.paper;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.UUID;

public class SafetyPolicy {
    private final Set<UUID> active = new LinkedHashSet<>();

    public synchronized void activate(UUID player) {
        active.add(player);
    }

    public synchronized void deactivate(UUID player) {
        active.remove(player);
    }

    public synchronized boolean isActive(UUID player) {
        return active.contains(player);
    }

    public synchronized boolean blocks(UUID player, SafetyAction action) {
        return blocks(player, action, false);
    }

    public synchronized boolean blocks(UUID player, SafetyAction action, boolean authorized) {
        return active.contains(player) && !authorized;
    }
}

File: src/main/kotlin/dev/jlo/gamemodes/paper/SafetyPolicy.kt
package dev.jlo.gamemodes.paper

import java.util.UUID

/** Safety actions that are never allowed for active participants without a plugin authorization token. */
enum class SafetyAction {
    BLOCK_BREAK, BLOCK_PLACE, BUCKET, FIRE, TNT, EXPLOSION, PISTON, HOPPER,
    PORTAL, BED, PEARL, CHORUS, VEHICLE, TELEPORT
}

class SafetyPolicy {
    private val active = linkedSetOf<UUID>()

    @Synchronized
    fun activate(player: UUID) { active += player }

    @Synchronized
    fun deactivate(player: UUID) { active -= player }

    @Synchronized
    fun isActive(player: UUID): Boolean = player in active

    @Synchronized
    fun blocks(player: UUID, action: SafetyAction, authorized: Boolean = false): Boolean =
        player in active && !authorized
}
